### PR TITLE
API: OpenAPI 3 spec and auto generated docs

### PR DESCRIPTION
**Pull request by rhymes**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [x] Bug Fix
- [ ] Documentation Update

## Description

This PR adds an (incomplete) [OpenAPI 3](http://spec.openapis.org/oas/v3.0.2) spec to the code base and the tooling to auto generate the doc.

OpenAPI 3 is a standard spec to describe REST APIs (formerly known as Swagger), there's is an [ecosystem of tooling](https://openapi.tools/) around it, from code generators, to validators, to extensions for IDEs to documentation generators.

I've added a spec for the Articles API (all the others are to be added) and removed some peculiarities that I don't think have been in use for a long time (see my review).

The auto generated doc uses [ReDoc](https://github.com/Redocly/redoc) which is a tool that auto generates a static and self contained HTML file. The styling, colors and so on [can be customized](https://github.com/Redocly/redoc/blob/master/src/theme.ts) by providing a custom theme if we want to use "company colors" instead of the default ones in the future.

A good example of a customized version is the Discourse API which is generated with the same tooling: https://docs.discourse.org/

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)

You can see the live example deployed by Netlify: https://5d4ab18d9ef1ac0008a9825b--devto.netlify.com/api/

Open this in a new tab so that you can zoom in:

Screenshot_2019-08-07 DEV API (beta)

## Added to documentation?

- [x] docs.dev.to
- [ ] readme
- [ ] no documentation needed

### CODE REVIEW

File: app/views/api/v0/articles/show.json.jbuilder [hunk 1]
```diff
@@ -22,8 +22,6 @@ json.last_comment_at    @article.last_comment_at&.utc&.iso8601
 
 json.body_html @article.processed_html
 json.body_markdown @article.body_markdown
-json.ltag_style(@article.liquid_tags_used.map { |ltag| Rails.application.assets["ltags/#{ltag}.css"].to_s.html_safe })
```
Comment (rhymes) [inline app/views/api/v0/articles/show.json.jbuilder]: these two properties, `ltag_style` and `ltag_script` are never used anywhere in the code either, they are neither documented. I think it's safe to remove them as well.

File: config/routes.rb [hunk 2]
```diff
@@ -241,7 +241,7 @@
 
   # You can have the root of your site routed with "root
   get "/about" => "pages#about"
-  get "/api", to: "pages#api"
+  get "/api", to: redirect("https://docs.dev.to/api")
```
Comment (rhymes) [inline config/routes.rb]: the existing "colloquial" API page has been replaced by the auto generated doc (which supports markdown in the descriptions by the way)

File: docs/contributing_api.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
+---
```
Comment (rhymes) [inline docs/contributing_api.md]: Added this to explain how to contribute to the API docs

File: app/controllers/api/v0/articles_controller.rb [hunk 4]
```diff
@@ -31,12 +31,7 @@ def index
       end
 
       def show
-        relation = Article.published.includes(:user)
-        @article = if params[:id] == "by_path"
-                     relation.find_by!(path: params[:url]).decorate
-                   else
-                     relation.find(params[:id]).decorate
-                   end
+        @article = Article.published.includes(:user).find(params[:id]).decorate
```
[the remark this answers is not in the document]
Comment (maestromac) [reply app/controllers/api/v0/articles_controller.rb]: looks like it should be safe.

File: app/views/api/v0/articles/show.json.jbuilder (hunk 1 above)
Comment (maestromac) [reply app/views/api/v0/articles/show.json.jbuilder]: nice catch!

**Comment by rhymes** [comment]
@maestromac 

> Love it! Really good looking API site. I wasn't aware that we our API endpoint is at ver 0.5.0

It's just an arbitrary version I used :D 

> Question regarding contributing on other editors, is it mostly manual editing api_v0.yml?

Yeah, I added the dev docs about how to do it with VSCode but it's mostly manual and it requires a bit of knowledge of the spec. There might be visual editors out there but I'm not aware of them.

**Comment by maestromac** [comment]
@rhymes do you plan on updating the docs with the latest changes or should I go ahead and merge this first?

**Comment by rhymes** [comment]
@maestromac could you merge this? I just started working on updating it to OAuth and the new endpoints in a separate branch

**Pull request merged**